packages/plugins/experimental/plugin-calls/src/state/call-manager.ts

```diff
@@ -0,0 +1,184 @@
+  }
+
+  turnAudioOff() {
+    return this._mediaManager.turnAudioOff();
+  }
+
+  turnVideoOn() {
+    return this._mediaManager.turnVideoOn();
+  }
+
+  turnVideoOff() {
+    return this._mediaManager.turnVideoOff();
+  }
+
+  turnScreenshareOn() {
+    return this._mediaManager.turnScreenshareOn();
+  }
+
+  turnScreenshareOff() {
+    return this._mediaManager.turnScreenshareOff();
+  }
+
+  constructor(private readonly _client: Client) {
+    super();
+    this._swarmSynchronizer = new CallSwarmSynchronizer({ networkService: _client.services.services.NetworkService! });
+    this._mediaManager = new MediaManager();
+  }
+
+  protected override async _open() {
+    await this._mediaManager.open();
+    await this._swarmSynchronizer.open();
+    const subscription = this._client.halo.identity.subscribe((identity) => {
+      if (identity) {
+        this._swarmSynchronizer._setIdentity(identity);
+      }
+      if (this._client.halo.device) {
+        this._swarmSynchronizer._setDevice(this._client.halo.device);
+      }
+    });
+
+    this._swarmSynchronizer.stateUpdated.on(this._ctx, (state) => this._onCallStateUpdated(state));
+    this._mediaManager.stateUpdated.on(this._ctx, (state) => this._onMediaStateUpdated(state));
+    this._ctx.onDispose(() => subscription.unsubscribe());
+  }
+
+  protected override async _close() {
+    await this._swarmSynchronizer.leave();
+    await this._swarmSynchronizer.close();
+    await this._mediaManager.close();
+  }
+
+  // TODO(mykola): Reconcile with _swarmSynchronizer.state.joined.
+  async join() {
+    this._swarmSynchronizer.setJoined(true);
+    await this._swarmSynchronizer.join();
+    await this._mediaManager.join({
+      iceServers: this._client.config.get('runtime.services.ice'),
+      apiBase: `${CALLS_URL}/api/calls`,
+    });
+  }
+
+  async leave() {
+    await this._swarmSynchronizer.leave();
+    this._swarmSynchronizer.setJoined(false);
+    await this._mediaManager.leave();
+  }
+
+  private _onCallStateUpdated(state: CallState) {
+    const tracksToPull = state.users
+      ?.filter((user) => user.joined && user.id !== state.self?.id)
+      ?.flatMap((user) => [user.tracks?.video, user.tracks?.audio, user.tracks?.screenshare])
+      .filter(isNonNullable);
+    this._mediaManager._schedulePullTracks(tracksToPull as EncodedTrackName[]);
+
+    this._updateState();
+  }
+
+  private _onMediaStateUpdated(state: MediaState) {
+    this._swarmSynchronizer.setTracks({
+      video: state.pushedVideoTrack ? TrackNameCodec.encode(state.pushedVideoTrack) : undefined,
+      audio: state.pushedAudioTrack ? TrackNameCodec.encode(state.pushedAudioTrack) : undefined,
+      screenshare: state.pushedScreenshareTrack ? TrackNameCodec.encode(state.pushedScreenshareTrack) : undefined,
+    });
+
+    this._updateState();
+  }
+
+  /**
+   * Only this method is allowed to change state.
+   */
+  private _updateState() {
+    this._state.call = this._swarmSynchronizer._getState();
```

Comment:
something to note here is that because this is overwriting the whole object it would blow away some reactivity, if it's working now it's probably fine but something to be aware of in the future as we look at adding more granular reactivity